Support galaxy v3/autohub API in ansible-galaxy

alikins opened the pull request:
##### SUMMARY

Make ansible-galaxy v3 aware, so that it can publish and download to Automation Hub. Not sure if we'll need to do anything special as far as authentication.

Fixes ansible/galaxy-dev#60


##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Feature Pull Request

##### COMPONENT NAME
ansible-galaxy

alikins committed:
Add galaxy collections API v3 support

Issue: ansible/galaxy-dev#60

- Determine if server supports v3

Use 'available_versions' from `GET /api`
to determine if 'v3' api is available on
the server.

- Support v3 pagination style

ie, 'limit/offset style', with the paginated
responses based on https://jsonapi.org/format/#fetching-pagination

v2 galaxy uses pagination that is more or less
'django rest framework style' or 'page/page_size style',
based on the default drf pagination described
at https://www.django-rest-framework.org/api-guide/pagination/#pagenumberpagination

- Support galaxy v3 style error response

The error objects returned by the galaxy v3 api are based
on the JSONAPI response/errors format
(https://jsonapi.org/format/#errors).

This handles that style response. At least for publish_collection
for now. Needs extracting/generalizing.

Handle HTTPError in CollectionRequirement.from_name()
with _handle_http_error(). It will raise AnsibleError
based on the json in an error response.

- Update unit tests

update test/unit/galaxy/test_collection*
to paramaterize calls to test against
mocked v2 and v3 servers apis.

Update artifacts_versions_json() to tale an
api version paramater.

Add error_json() for generating v3/v3 style error
responses.

So now, the urls generated and the pagination schema
of the response will use the v3 version if
the passed in GalaxyAPI 'galaxy_api' instance
has 'v3' in it's available_api_versions

alikins commented:
Unit tests only cover the case where there is no /api/v3 at the moment, so this is not very well tested currently.

alikins committed:
Move checking of server avail versions to collections.py

collections.py needs to know the server api versions
supported before it makes collection related calls,
so the 'lazy' server version check in api.GalaxyAPI
is never called and isn't set, so 'v3' servers weren't
found.

Update unit tests to mock the return value of the
request instead of GalaxyAPI itself.

chouseknecht commented on lib/ansible/galaxy/api.py:
```diff
@@ -43,11 +43,14 @@ def wrapped(self, *args, **kwargs):
             server_version = self._get_server_api_version()
+
             if server_version not in self.SUPPORTED_VERSIONS:
                 raise AnsibleError("Unsupported Galaxy server API version: %s" % server_version)
```
We said that we're going to deprecate current_version from the Galaxy API, so it seems we should remove this check altogether and just default to `v1`.

ironfroggy replied:
We can't actually get rid of it, since existing supported versions of `ansible-galaxy` depend on it. But we _should_ remove uses of it and rely on `available_versions` instead and make sure there's an overlap between that and `SUPPORTED_VERSIONS` here.

alikins replied:
Wasn't planning on changing _get_server_api_version() related code for this PR given the time frame, but can revisit if necessary.

chouseknecht replied:
Let it go for now.

chouseknecht reviewed (approved):
One minor comment regarding removing use of `current_version`.

ironfroggy commented on lib/ansible/galaxy/collection.py:
```diff
@@ -356,6 +375,29 @@ def from_name(collection, apis, requirement, force, parent=None):
+
+    available_versions = data.get('available_versions',
+                                  {'v1': '/api/v1',
+                                   'v2': '/api/v2'})
```
why does this have a default value? don't we only want to use what the server says and never guess?

alikins replied:
Can't use what the server says because the live deployed galaxy.ansible.com doesn't include 'v2'.
Supposedly because '/api/v2' is not "public" "stable" API. I disagree with that reasoning, but nonetheless.

``` json
{
    "description": "GALAXY REST API",
    "current_version": "v1",
    "available_versions": {
        "v1": "/api/v1/"
    },
    "server_version": "3.2.4",
    "version_name": "Doin' it Right",
    "team_members": [
        "chouseknecht",
        "cutwater",
        "alikins",
        "newswangerd",
        "awcrosby",
        "tima",
        "gregdek"
    ]
}
```

ironfroggy replied:
But you _are_ using the server response here, you're just providing a default if its missing. `galaxy.ansible.com` is missing V2, but it still provides a `available_versions` that (currently) only includes V1, so this V1/V2 dict will never actually be used.

ironfroggy commented on lib/ansible/galaxy/collection.py:
```diff
@@ -409,28 +451,25 @@ def publish_collection(collection_path, api, wait, timeout):
+    available_api_versions = get_available_api_versions(api)
+
+    if 'v3' in available_api_versions:
+        n_url = _urljoin(api.api_server, 'api', 'v3', 'artifacts', 'collections')
```
should the other version of `n_url` be in an `else` block?

alikins replied:
It could be, but I prefer the 'set the default, but override for the special case' style.

ironfroggy commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1016,3 +1055,33 @@ def _extract_tar_file(tar, filename, b_dest, b_temp_path, expected_hash=None):
         shutil.move(to_bytes(tmpfile_obj.name, errors='surrogate_or_strict'), b_dest_filepath)
+
+
+def _handle_http_error(http_error, api, available_api_versions, context_error_message):
```
I think this deserves a docstring noting for future maintainers that this can't be used outside of collection endpoints, because it won't handle V1 error responses. Someone might get the idea to use this elsewhere and easily overlook that constraint.

jborean93 replied:
Probably better if it was changed to something like `handle_galaxy_error` or something.

alikins replied:
@ironfroggy More docs are better of course, but this is dunder and in collections.py because it only handles v2/v3 style responses.

ironfroggy replied:
fair enough, tho i'm with @jborean93 on the naming

ironfroggy commented on test/units/galaxy/test_collection_install.py:
```diff
@@ -56,20 +56,57 @@ def artifact_json(namespace, name, version, dependencies, server):
+            response = errors_to_return[0]
+
+    if api_version == 'v3':
+        response['errors'] = errors_to_return
```
Does this make sense up with first v3 block rather than splitting the logic in half?

alikins replied:
Could be either way. Is the logic incorrect here?

ironfroggy replied:
Its not incorrect, just harder to read.

alikins committed:
Handle 'GET /api' requiring auth for v3

Try unauth `GET /API`, and if we get a 401,
then try the v3 style Bearer token auth.

alikins committed:
Set GalaxyAPI inst 'token_type' if we know it's api version

In a few places, already instantiated GalaxyAPI api objects
will get there 'token_type' modifed and set when we figure
out what API version they are.

Galaxy API v2 uses 'Token' and v3 uses 'Bearer' when
constructing the http 'Authorization' header with the
api key.

alikins committed:
Default to trying `GET /api` with auth if available

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -310,6 +310,14 @@ def from_name(collection, apis, requirement, force, parent=None):
         for api in apis:
             collection_url_paths = [api.api_server, 'api', 'v2', 'collections', namespace, name, 'versions']
+
+            available_api_versions = get_available_api_versions(api)
```
This is can be a pretty expensive operation, especially when we have multiple collections that we are trying to install. We should be setting the available versions once on the `api` object so that once we only need to do this check once for every server.

alikins replied:
@jborean93 The tradeoff here is the cases where the API version needs to be known before the call is made. 

If those calls can be moved into GalaxyAPI and wrapped with gconnect() and abstracted slightly, should be able to ensure the `GET /api` is only called once per GalaxyAPI instance.

The branch as is at the moment was focused on minimizing changes to `ansible.galaxy.api` and isolating the changes to collection.py. But sounds like I should change that focus?

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -310,6 +310,14 @@ def from_name(collection, apis, requirement, force, parent=None):
+            available_api_versions = get_available_api_versions(api)
+            if 'v3' in available_api_versions:
+                # /api/v3/ exists, use it
+                collection_url_paths[2] = 'v3'
```
It sounds like this should be set on the api object and used as part of the original url paths

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -310,6 +310,14 @@ def from_name(collection, apis, requirement, force, parent=None):
+                # /api/v3/ exists, use it
+                collection_url_paths[2] = 'v3'
+                # update this v3 GalaxyAPI to use Bearer token from now on
+                api.token_type = 'Bearer'
```
Does it make sense for `Bearer` to be set in `_auth_header()` based on the version used instead of you manually setting it each time for v3 endpoints.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -356,6 +378,45 @@ def from_name(collection, apis, requirement, force, parent=None):
         return req
 
 
+def get_available_api_versions(galaxy_api):
```
As I mentioned above, this seems like it should be part of the `GalaxyAPI` class and not here. You could make it so it only collections the versions when someone tries to access the `available_api_version` property lazily to avoid running it multiple times on the same server.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -356,6 +378,45 @@ def from_name(collection, apis, requirement, force, parent=None):
+    headers = {}
+    headers.update(galaxy_api._auth_header(required=False))
+
+    url = _urljoin(galaxy_api.api_server, "api")
```
You need to fix up your API endpoints to actually return `v2`, right now all of these endpoints (dev, qa, prod) only return `v1` https://galaxy.ansible.com/api/.

alikins replied:
The 'current_version' from `GET /api` (as used in GalaxyAPI.version and GalaxyAPI.baseurl) isn't really meaningful anymore. 

Existing clients will break if it is not 'v1'. ansible-galaxy talking to galaxy.ansible.com ends up needing both the 'v1' urls (roles, etc) and the 'v2' urls (collections, artifacts). ie, 'v2' api is not a superset of 'v1', so both are used. And for talking to automation hub, it needs just 'v3'.

The plan is to eventually drop 'current_version' from `GET /api` results, but we can't change it for production galaxy.ansible.com without breaking deployed ansible-galaxy versions.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -356,6 +378,45 @@ def from_name(collection, apis, requirement, force, parent=None):
 
+def get_available_api_versions(galaxy_api):
+    headers = {}
+    headers.update(galaxy_api._auth_header(required=False))
```
you could just do `headers = galaxy_api._auth_header(required=False)`, not much need for .update here

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -410,27 +471,25 @@ def publish_collection(collection_path, api, wait, timeout):
     display.display("Publishing collection artifact '%s' to %s %s" % (collection_path, api.name, api.api_server))
 
     n_url = _urljoin(api.api_server, 'api', 'v2', 'collections')
+    available_api_versions = get_available_api_versions(api)
```
Ditto here, available_versions should be a property on the api instance which is lazily collected as needed.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -410,27 +471,25 @@ def publish_collection(collection_path, api, wait, timeout):
 
     data, content_type = _get_mime_data(b_collection_path)
-    headers = {
+    headers.update({
```
Is there a reason for changing the way the headers are set. I don't really care too much but it just adds more noise to the PR for no reason that I can see.

alikins replied:
Hmm, probably no reason to change it. I must of changed it in previous version at some point. But I'll change it back and verify.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1016,3 +1075,33 @@ def _extract_tar_file(tar, filename, b_dest, b_temp_path, expected_hash=None):
+    except (AttributeError, ValueError):
+        err_info = {}
+
+    if 'v3' in available_api_versions:
```
this should be sourced from the request/response URL as that's where the error is from.

alikins replied:
something like

``` python
# or more likely, with regex and/or urlparse(url)[2] and/or inspect the response content
if '/api/v3' in http_error.geturl():
   ....
```

(In theory, the urls may not have 'v3' in them, but we make that assumption elsewhere already so should be ok).

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -1016,3 +1075,33 @@ def _extract_tar_file(tar, filename, b_dest, b_temp_path, expected_hash=None):
+            errors = [{'detail': 'Unknown error returned by Galaxy server.',
+                       'code': 'Unknown'}]
+
+        for error in errors:
```
if `errors` is not in the response then `errors` is `None` which will fail here.

alikins replied:
Not sure I understand, if the response doesn't have an 'errors' field, errors will be None at line 1090 and will be set to the list with the fallback/default error.

jborean93 commented on lib/ansible/galaxy/api.py:
```diff
@@ -63,25 +66,32 @@ class GalaxyAPI(object):
         self.username = username
         self.password = password
         self.token = token
+        self.token_type = token_type or 'Token'
```
It seems like this should potentially be based on the API version that is being used and not something explicitly set. Would there be a reason why this should be configurable outside of using `Bearer` for `v3` and `Token` for older versions?

alikins replied:
Considered passing in api_version or some slightly more abstract authentication option and letting GalaxyAPI choose the right auth mechanism based on that. 

Chose to pass in 'token_type' since it seemed more consistent with the existing params which are all very explicit and concrete. Though, at the moment, nothing passes 'token_type' to the the GalaxyAPI() init (ie, it's not configurable or setup from cli/galaxy.py for ex). So the token_type param could be removed and default to 'Token'[1]

I'm taking a look at what changes to use api_version would look like. Assuming no problems I can change it.

[1] or api_version or auth_type etc defaults could be setup.

jborean93 commented on lib/ansible/galaxy/api.py:
```diff
@@ -63,25 +66,32 @@ class GalaxyAPI(object):
         self.baseurl = None
         self.version = None
         self.initialized = False
+        self.available_api_versions = {}
```
Leading into my other comments, I think this should be lazily evaluated on this class when called for the first time to save the time it takes for `ansible-galaxy` to complete an operation.

alikins replied:
@jborean93 The problem I ran into trying to set it up lazily was a bit of a catch-22 for things like 'publish_collection' where the API version needs to be known before the `PUSH /api/v2/collections/` or `PUSH /api/v3/artifacts/collections/` is made so the right url is chosen. So I changed it to explicitly check the avail versions before those calls.

I could move those calls into the GalaxyAPI class and wrap them with @gconnect(). And abstract them slightly for v2/v3 diffs. 

Any reason not to move those url requests into GalaxyAPI? I left the galaxy API open_urls() calls in collection.py, Should they be moved into ansible.galaxy.api.GalaxyAPI() ?

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -356,6 +378,45 @@ def from_name(collection, apis, requirement, force, parent=None):
+        raise AnsibleError("Failed to get data from the API server (%s): %s " % (url, to_native(e)))
+
+    try:
+        data = json.loads(to_text(return_data.read(), errors='surrogate_or_strict'))
```
if `return_data` is the response from `open_url`, you should be able to do `json.load(return_data)` and not worry about reading from the io reader and converting to text

jborean93 reviewed (changes requested):
Most stuff looks fine, would like to see the get available APIs to be more optimized and only collected as needed and once per server.

jborean93 commented:
@alikins let's get this in a form so we can merge then look at optimising things afterwards. The branch you shared does solve a few of the things I talked about here but because of the deadlines I prefer get something in that doesn't break the existing behaviour and then fix later

alikins committed:
pep8 fix for test_collection.py

alikins committed:
Fix py27 test vailure (TypeError on StringIO)

jborean93 commented:
Draft PR that implements some of the changes talked about here https://github.com/ansible/ansible/pull/61510.
